packages/permission-controller/src/PermissionController.ts

```diff
@@ -1991,6 +2110,214 @@ export class PermissionController<
+
+        Object.entries(incrementalRequestedPermissions).forEach(
+          ([targetName, rightPermission]) => {
+            if (
+              !hasProperty(
+                // We type-guard the targetName key out of existence without this cast.
+                leftPermissions as Record<string, unknown>,
+                targetName,
+              )
+            ) {
+              const [newPermission, caveatsDiff] = this.#mergePermission(
+                {},
+                rightPermission,
+              );
+
+              leftPermissions[targetName] = newPermission;
+              permissionDiffMap[targetName] = caveatsDiff;
+            } else {
+              const [newPermission, caveatsDiff] = this.#mergePermission(
+                leftPermissions[targetName],
+                rightPermission,
+              );
+
+              if (Object.keys(caveatsDiff).length > 0) {
+                leftPermissions[targetName] = newPermission;
+                permissionDiffMap[targetName] = caveatsDiff;
+              }
+              // Otherwise, leave the left permission as-is; its authority has
+              // not changed.
+            }
+          },
+        );
+      },
+    );
+
+    if (Object.keys(permissionDiffMap).length === 0) {
+      return [];
+    }
+    return [newPermissions, permissionDiffMap];
+  }
+
+  /**
+   * Performs a right-biased union between two permissions. The task of merging caveats
+   * of the same type between the two permissions is delegated to the corresponding
+   * caveat type's merger implementation.
+   *
+   * Throws if the left-hand and right-hand permissions both have a caveat whose
+   * specification does not provide a caveat merger function.
+   *
+   * @param leftPermission - The left-hand permission to merge.
+   * @param rightPermission - The right-hand permission to merge.
+   * @returns The merged permission.
+   */
+  #mergePermission(
+    leftPermission: Partial<PermissionConstraint>,
+    rightPermission: Partial<PermissionConstraint>,
+  ): [Partial<PermissionConstraint>, CaveatDiffMap<CaveatConstraint>] {
+    const { caveatPairs, leftUniqueCaveats, rightUniqueCaveats } =
+      this.#collectUniqueAndPairedCaveats(leftPermission, rightPermission);
+
+    const [mergedCaveats, caveatDiffMap] = caveatPairs.reduce(
+      ([caveats, diffMap], [leftCaveat, rightCaveat]) => {
+        const [newCaveat, diff] = this.#mergeCaveat(leftCaveat, rightCaveat);
+
+        if (newCaveat !== undefined && diff !== undefined) {
+          caveats.push(newCaveat);
+          diffMap[newCaveat.type] = diff;
+        } else {
+          caveats.push(leftCaveat);
+        }
+
+        return [caveats, diffMap];
+      },
+      [[], {}] as [CaveatConstraint[], CaveatDiffMap<CaveatConstraint>],
+    );
+
+    const mergedRightUniqueCaveats = rightUniqueCaveats.map((caveat) => {
+      const [newCaveat, diff] = this.#mergeCaveat(undefined, caveat);
```

Comment:
Couldn't we just add the right unique caveats as-is without requiring the merging to support an undefined left side? Potentially simplifying the merger function downstream

Reply:
It would be nice, but I don't think it's possible at the moment because of the diff.

The format of the diff is ultimately left as an exercise to the caveat implementer. Therefore, if we naively use the right-hand value as the diff when the left-hand side is `undefined`, we run the risk of producing differently formatted diffs for the same caveat type.

Now, `ARCHITECTURE.md` _does_ specify that the diff should be "expressed in the type of the merged values", and this soft requirement could be workable in practice. However, I'm also worried that it could cause trouble down the road.

As an aside, once (if) we have a `superstruct` DSL for specifying caveats, all of these issues go away.

What do you think?

Reply:
I decided to let the controller handle undefined left-hand sides in `#mergeCaveat()`. See: [48ad5ea](https://github.com/MetaMask/core/pull/4222/commits/48ad5eaa02653ad9044fe87d3264ada341fc63c8)